drivers/adc/adc_stm32f4/src/adc_stm32f4.c

```diff
@@ -0,0 +1,712 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+#include <hal/hal_bsp.h>
+#include <adc/adc.h>
+#include <assert.h>
+#include <os/os.h>
+#include <bsp/cmsis_nvic.h>
+#include "stm32f4xx_hal_dma.h"
+#include "stm32f4xx_hal_adc.h"
+#include "stm32f4xx_hal_rcc.h"
+#include "stm32f4xx_hal_cortex.h"
+#include "stm32f4xx_hal.h"
+#include "adc_stm32f4/adc_stm32f4.h"
+#include "stm32f4xx_hal_dma.h"
+#include "mcu/stm32f4xx_mynewt_hal.h"
+
+#define STM32F4_IS_DMA_ADC_CHANNEL(CHANNEL) ((CHANNEL) <= DMA_CHANNEL_2)
+
+static DMA_HandleTypeDef *dma_handle[5];
+static struct adc_dev *adc_dma[5];
+
+struct stm32f4_adc_stats {
+    uint16_t adc_events;
+    uint16_t adc_error;
+    uint16_t adc_DMA_xfer_failed;
+    uint16_t adc_DMA_xfer_aborted;
+    uint16_t adc_DMA_xfer_complete;
+    uint16_t adc_DMA_start_error;
+    uint16_t adc_DMA_overrun;
+    uint16_t adc_internal_error;
+};
+
+static struct stm32f4_adc_stats stm32f4_adc_stats;
+
+static void
+stm32f4_adc_clk_enable(ADC_HandleTypeDef *hadc) {
+    uintptr_t adc_addr = (uintptr_t)hadc->Instance;
+
+    switch (adc_addr) {
+        case (uintptr_t)ADC1:
+            __HAL_RCC_ADC1_CLK_ENABLE();
+            break;
+        case (uintptr_t)ADC2:
+            __HAL_RCC_ADC2_CLK_ENABLE();
+            break;
+        case (uintptr_t)ADC3:
+            __HAL_RCC_ADC3_CLK_ENABLE();
+            break;
+        default:
+            assert(0);
+    }
+}
+
+static void
+stm32f4_adc_clk_disable(ADC_HandleTypeDef *hadc) {
+    uintptr_t adc_addr = (uintptr_t)hadc->Instance;
+
+    switch (adc_addr) {
+        case (uintptr_t)ADC1:
+            __HAL_RCC_ADC1_CLK_DISABLE();
+            break;
+        case (uintptr_t)ADC2:
+            __HAL_RCC_ADC2_CLK_DISABLE();
+            break;
+        case (uintptr_t)ADC3:
+            __HAL_RCC_ADC3_CLK_DISABLE();
+            break;
+        default:
+            assert(0);
+    }
+}
+
+static GPIO_InitTypeDef
+stm32f4_resolve_adc_gpio(ADC_HandleTypeDef *adc, uint8_t cnum)
+{
+    uintptr_t adc_addr = (uintptr_t)adc->Instance;
+    uint32_t pin;
+    GPIO_InitTypeDef gpio_td;
+
+    switch (adc_addr) {
+        case (uintptr_t)ADC1:
+        case (uintptr_t)ADC2:
```

Comment:
ADC_1 and ADC_2 share the pins and hence the switches fall through. Also, similar is the case for ADC_1, ADC_2 and ADC_3.